src/groups/mqb/mqbblp/mqbblp_queueconsumptionmonitor.cpp

```diff
@@ -199,99 +200,372 @@ void QueueConsumptionMonitor::reset()
+            &subStreamInfo->d_idleEventHandle);
+        subStreamInfo->d_idleEventHandle.release();
+    }
+
+    subStreamInfo->d_state = State::e_ALIVE;
+
+    bdlma::LocalSequentialAllocator<2048> localAllocator(0);
+
+    bmqt::UriBuilder uriBuilder(d_queueState_p->uri(), &localAllocator);
+    uriBuilder.setId(appId);
+
+    bmqt::Uri uri(&localAllocator);
+    uriBuilder.uri(&uri);
+
+    BALL_LOG_INFO << "Queue '" << uri << "' no longer appears to be stuck.";
+}
+
+void QueueConsumptionMonitor::scheduleAlarmEvent(
+    const bsls::TimeInterval& alarmTime)
+{
+    // executed by the *QUEUE DISPATCHER* thread
+
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_queueState_p->queue()->dispatcher()->inDispatcherThread(
+        d_queueState_p->queue()));
+    BSLS_ASSERT_SAFE(!d_alarmEventHandle);
+
+    d_scheduledAlarmTime = alarmTime;
+
+    d_queueState_p->scheduler()->scheduleEvent(
+        &d_alarmEventHandle,
+        d_scheduledAlarmTime,
+        bdlf::BindUtil::bind(
+            &QueueConsumptionMonitor::executeAlarmInQueueDispatcher,
+            this));
+}
+
+void QueueConsumptionMonitor::scheduleOrRescheduleAlarmEventIfNeeded(
+    const bsls::TimeInterval& alarmTime)
+{
+    // executed by the *QUEUE DISPATCHER* thread
+
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_queueState_p->queue()->dispatcher()->inDispatcherThread(
+        d_queueState_p->queue()));
+
+    if (d_alarmEventHandle) {
+        // If the event is already scheduled, reschedule it only if the new
+        // alarm time is earlier than the currently scheduled one.
+        if (alarmTime < d_scheduledAlarmTime) {
+            d_scheduledAlarmTime = alarmTime;
+            d_queueState_p->scheduler()->rescheduleEvent(d_alarmEventHandle,
+                                                         d_scheduledAlarmTime);
+        }
+    }
+    else {
+        // If the event is not scheduled, schedule it.
+        scheduleAlarmEvent(alarmTime);
+    }
+}
+
+void QueueConsumptionMonitor::scheduleIdleEvent(SubStreamInfo* subStreamInfo,
+                                                const bsl::string& appId)
+{
+    // executed by the *QUEUE DISPATCHER* thread
+
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_queueState_p->queue()->dispatcher()->inDispatcherThread(
+        d_queueState_p->queue()));
+    BSLS_ASSERT_SAFE(!subStreamInfo->d_idleEventHandle);
+    BSLS_ASSERT_SAFE(subStreamInfo->d_state == State::e_IDLE);
+
+    bsls::TimeInterval idleTime = d_queueState_p->scheduler()->now();
+    idleTime.addSeconds(bsl::min(d_maxIdleTimeSec, k_IDLE_TIMER_PERIOD_SEC));
+    d_queueState_p->scheduler()->scheduleEvent(
+        &subStreamInfo->d_idleEventHandle,
+        idleTime,
+        bdlf::BindUtil::bind(
+            &QueueConsumptionMonitor::executeIdleInQueueDispatcher,
+            this,
+            appId));
+}
+
+void QueueConsumptionMonitor::executeAlarmInQueueDispatcher()
+{
+    // executed by the *SCHEDULER DISPATCHER* thread
```

Comment:
```suggestion
    // executed by the *SCHEDULER* thread
```

Scheduler thread is not a dispatcher thread.